lib/src/rules/library_private_types_in_public_api.dart

```diff
@@ -0,0 +1,242 @@
+    if (Identifier.isPrivateName(node.name.name)) {
+      return;
+    }
+    node.returnType?.accept(this);
+    node.typeParameters?.accept(this);
+    node.parameters.accept(this);
+  }
+
+  @override
+  void visitFunctionTypedFormalParameter(FunctionTypedFormalParameter node) {
+    if (node.isNamed && Identifier.isPrivateName(node.identifier.name)) {
+      return;
+    }
+    node.returnType?.accept(this);
+    node.typeParameters?.accept(this);
+    node.parameters.accept(this);
+  }
+
+  @override
+  void visitGenericFunctionType(GenericFunctionType node) {
+    node.returnType?.accept(this);
+    node.typeParameters?.accept(this);
+    node.parameters.accept(this);
+  }
+
+  @override
+  void visitGenericTypeAlias(GenericTypeAlias node) {
+    if (Identifier.isPrivateName(node.name.name)) {
+      return;
+    }
+    node.typeParameters?.accept(this);
+    node.functionType?.accept(this);
+  }
+
+  @override
+  void visitMethodDeclaration(MethodDeclaration node) {
+    if (Identifier.isPrivateName(node.name.name)) {
+      return;
+    }
+    node.returnType?.accept(this);
+    node.typeParameters?.accept(this);
+    node.parameters?.accept(this);
+  }
+
+  @override
+  void visitMixinDeclaration(MixinDeclaration node) {
+    if (Identifier.isPrivateName(node.name.name)) {
+      return;
+    }
+    node.onClause?.superclassConstraints.accept(this);
+    node.typeParameters?.accept(this);
+    node.members.accept(this);
+  }
+
+  @override
+  void visitSimpleFormalParameter(SimpleFormalParameter node) {
+    var name = node.identifier;
+    if (name != null && node.isNamed && Identifier.isPrivateName(name.name)) {
+      return;
+    }
+    node.type?.accept(this);
+  }
+
+  @override
+  void visitTopLevelVariableDeclaration(TopLevelVariableDeclaration node) {
+    if (node.variables.variables
+        .any((field) => !Identifier.isPrivateName(field.name.name))) {
+      node.variables.type?.accept(this);
+    }
+  }
+
+  @override
+  void visitTypeArgumentList(TypeArgumentList node) {
+    node.arguments.accept(this);
+  }
+
+  @override
+  void visitTypeName(TypeName node) {
+    var element = node.name.staticElement;
+    if (element != null && isPrivate(element)) {
+      rule.reportLint(node.name);
+    }
+    node.typeArguments?.accept(this);
+  }
+
+  @override
+  void visitTypeParameter(TypeParameter node) {
+    node.bound?.accept(this);
+  }
+
+  @override
+  void visitTypeParameterList(TypeParameterList node) {
+    node.typeParameters.accept(this);
+  }
+
+  /// Return `true` if the given [element] is private or is defined in a private
+  /// library.
+  static bool isPrivate(Element element) {
+    var name = element.name;
+    return name != null && Identifier.isPrivateName(name);
+  }
+}
+
+class Visitor extends SimpleAstVisitor {
+  LintRule rule;
+
+  Visitor(this.rule);
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    if (!Validator.isPrivate(node.declaredElement!)) {
```

Comment:
We're sure `declaredElement` can't be null here because the AST is sure to be resolved?  Since it's hard to know this is safe from just looking, I might prefer:

```dart
var element = node.declaredElement;
if (element != null && !Validator.isPrivate(element)) {
   ...
}
```

but don't feel that strongly.

Reply:
Done.